cmd/request.go

```diff
@@ -0,0 +1,145 @@
+package cmd
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/tkhq/tkcli/internal/apikey"
+	"github.com/tkhq/tkcli/internal/clifs"
+	"github.com/tkhq/tkcli/internal/display"
+)
+
+var (
+	requestHost, requestPath, requestBody string
+	requestNoPost, requestShowCurlCommand bool
+)
+
+func init() {
+	makeRequest.Flags().StringVar(&requestHost, "host", "coordinator-beta.turnkey.io", "hostname of the API server")
+	makeRequest.Flags().StringVar(&requestPath, "path", "", "path for the API request")
+	makeRequest.Flags().StringVar(&requestBody, "body", "-", "body of the request, which can be '-' to indicate stdin or be prefixed with '@' to indicate a source filename")
+	makeRequest.Flags().BoolVar(&requestNoPost, "no-post", false, "only provide the signature, do not post the request to the API server")
+	makeRequest.Flags().BoolVar(&requestShowCurlCommand, "show-curl", false, "only provide the signature, do not post the request to the API server")
+
+	rootCmd.AddCommand(makeRequest)
+}
+
+var makeRequest = &cobra.Command{
+	Use:     "request takes a request body, generates a stamp for the given request and optionally sends it to the Turnkey API server",
+	Short:   "request makes a basic API request",
+	Aliases: []string{"req", "r"},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		protocol := "https"
+		if pattern := regexp.MustCompile(`^localhost:\d+$`); pattern.MatchString(requestHost) {
+			protocol = "http"
+		}
+
+		apiKey, err := clifs.LoadKeypair(KeyName)
+		if err != nil {
+			return errors.Wrap(err, "failed to get API key")
+		}
+
+		bodyReader, err := processRequestBody(requestBody)
+		if err != nil {
+			return errors.Wrap(err, "failed to process request body")
+		}
+
+		body, err := io.ReadAll(bodyReader)
+		if err != nil {
+			return errors.Wrap(err, "failed to read message body")
+		}
+
+		stamp, err := apikey.Signature(body, apiKey)
+		if err != nil {
+			return errors.Wrap(err, "failed to produce a valid API stamp")
+		}
+
+		enc := json.NewEncoder(cmd.OutOrStdout())
+		enc.SetIndent("", "   ")
+
+		if requestNoPost {
+			ret := map[string]string{
```

Comment:
These pain me greatly:  we should be using []byte here instead of string, but to keep compatibility, we have to use string, because json.Marshal helpfully base64-encodes []byte types.
That is/would be great, and we could throw out much of the key encoding stuff... but compatibility.